File: Algorithms/Python/Knuth–Morris–Pratt.py


def findPattern(text, pattern):
    '''
    Knuth–Morris–Pratt Algorithm
    '''
    if pattern == "": return 0
    
    # Defining Longest Prefix Suffix
    lps=[0]*len(pattern)
    prevLPS=0
    i=1
    while i<len(pattern):
        if pattern[i]==pattern[prevLPS]:
            lps[i]=prevLPS+1
            prevLPS += 1
            i += 1
        elif prevLPS==0:
            lps[i]=0
            i += 1
        else:
            prevLPS=lps[prevLPS-1]
            
    # Finding patterin in the text
    i=0 # text pointer
    j=0 # pattern pointer
    while i<len(text):
        if text[i]==pattern[j]:
            i += 1
            j += 1
        else:
            if j==0:
                i += 1
            else:
                j=lps[j-1]
            
        if j==len(pattern):
            return i-len(pattern)
    
    return -1

#         0123456789
source = "----abcxxxabcy---------";
patt = "abcxxxabcy";

index = findPattern(source, patt)
print("Found pattern: "+patt+"  on index: "+str(index)+" in text: "+source) # Found pattern: abcxxxabcy  on index: 4


#         01234567890    
source = "123456yobanaV346563sdsdkhfhs"
pattern = "yobana"
index = findPattern(source, pattern)
print("Found pattern: "+pattern+"  on index: "+str(index)+" in text: "+source) # Found pattern: abcxxxabcy  on index: 6

File: Algorithms/CPP/Knuth–Morris–Pratt.cpp
/*
g++ Knuth–Morris–Pratt.cpp -std=c++17 -o /tmp/a.out;/tmp/a.out
*/

#include <bits/stdc++.h>
using namespace std;

int findPattern(string text, string pattern) {
    /*
    Knuth–Morris–Pratt Algorithm
     28. Implement strStr()
    */
    if (pattern == "") return 0;
    
    // Setup Longest Prefix Suffix array
    vector<int>lps(pattern.size(), 0);
    int prevLPS = 0;
    int i = 1;
    
    while (i<pattern.size()) {
        if (pattern[i]==pattern[prevLPS]) {
            lps[i] = prevLPS + 1;
            prevLPS++;
            i++;
            
        } else if (prevLPS==0) {
            lps[i] = 0;
            i++;
        } else {
            prevLPS = lps[prevLPS-1];
        }
    }
    
    // Search pattern in the text
    i = 0; // pointer for the text
    int j = 0; // pointer for the pattern
    while (i<text.size()) {
        if (text[i]==pattern[j]) {
            i++;
            j++;
        } else {
            if (j==0) {
                i++;
            } else {
                j=lps[j-1];
            }
        }
        if (j==pattern.size()) 
            return i-pattern.size();
    }
    return -1;
}

int main() {
    cout << endl << "-----------------------------Knuth–Morris–Pratt" << endl; 
    string patt = "abcxxxabcy";
    string text = "----abcxxxabcy---------";
    int index = findPattern(text, patt);
    if (index != -1) {
        cout << "Pattern: "<< patt << " found at: `" << index << "`" <<  " in text: `" << text << "`" << "\n"; // Pattern: abcxxxabcy found at: `4` in text: `----abcxxxabcy---------`
    } else {
        cout << "Pattern: "<< patt << " not found in text: `" << text << "`" <<endl;
    }

    return 0;
}